### ISSUE TITLE
Make it easier to subclass MoyaProvider

### ISSUE THREAD

**Issue by ashfurrow** [Moya/Moya#144]
It uses default values for its initializers that need to be duplicated by subclasses. 

```swift
public init(endpointClosure: EndpointClosure = MoyaProvider.DefaultEndpointMapping,
    requestClosure: RequestClosure = MoyaProvider.DefaultRequestMapping,
    stubClosure: StubClosure = MoyaProvider.NeverStub,
    networkActivityClosure: Moya.NetworkActivityClosure? = nil,
    credentialClosure: CredentialClosure? = nil,
    manager: Manager = Alamofire.Manager.sharedInstance) {

    self.endpointClosure = endpointClosure
    self.requestClosure = requestClosure
    self.stubClosure = stubClosure
    self.networkActivityClosure = networkActivityClosure
    self.credentialClosure = credentialClosure
    self.manager = manager
}
```

Dunno how we could do this..

**Comment by ashfurrow** [comment]
This is really tricky. Basically our providers are containers for data (their configs) and a `request` method. That's, like, the definition of a class. 

I suppose we could move `MoyaProvider` to be a protocol and put `request()` in a protocol extension? Or maybe turn Moya into a `struct` instead of a class. The real hassle is just updating initializers that don't do anything but call `super.init(...` so structs would remove that pain most of the time.

**Comment by ashfurrow** [comment]
Alternatively, it seems a bit odd that subclasses _must_ override the initializer. 

```swift
override public init(endpointClosure: EndpointClosure = MoyaProvider.DefaultEndpointMapping,
    requestClosure: RequestClosure = MoyaProvider.DefaultRequestMapping,
    stubClosure: StubClosure = MoyaProvider.NeverStub,
    networkActivityClosure: Moya.NetworkActivityClosure? = nil,
    credentialClosure: CredentialClosure? = nil,
    manager: Manager = Alamofire.Manager.sharedInstance) {
    
        super.init(endpointClosure: endpointClosure, requestClosure: requestClosure, stubClosure: stubClosure, networkActivityClosure: networkActivityClosure, credentialClosure: credentialClosure, manager: manager)
}
```

All the subclasses ever do is call `super.init()`. How can we abstract this?

**Comment by ashfurrow** [comment]
Maybe instead of subclassing, we can add extensions to the existing `MoyaProvider` type?

**Comment by ashfurrow** [comment]
I like the idea of using protocols to extend the primary type, but I'm concerned about the constraints that would introduce for existing users who are currently subclassing (say, `RxSwiftProvider`). I'm going to look through [Eidolon](http://github.com/artsy/eidolon) for how we use Moya, and see if there are any foreseeable problems before starting an implementation.

**Comment by ashfurrow** [comment]
Fixed in #238.

**Comment by nikita-leonov** [comment]
Was it actually addressed? We are trying to subclass both `MoyaProvider` and `ReactiveCocoaMoyaProvider` and Xcode says we should override designated initializer.

**Comment by ashfurrow** [comment]
Well, it was addressed in the sense that subclassing the provider no longer requires so many initializer parameters and repeating their default values. What you're describing sounds like a new issue. I have subclassed `RxMoyaProvider` [here](https://github.com/artsy/eidolon/blob/1c746f9a9c6420cda9f691468783fee67376e8b1/Kiosk/App/Networking/Networking.swift#L7) without issues.

**Comment by nikita-leonov** [comment]
I see. Yes, it is easier to define default parameters now, but it still require overriding designated initializer. Actually situation does not change in compare with what you writing in Oct, 1. Here is current experience of subclassing — https://www.evernote.com/l/AEaV5Z1HlZRGprguAUiszmTarcbrRgFzQA0

**Comment by ashfurrow** [comment]
Subclassing in Swift usually requires calling a designated initializer in the superclass, I'm not sure of a great way to avoid this in Moya. Do you have suggestions?  

--  
Ash Furrow
https://ashfurrow.com/

**Comment by nikita-leonov** [comment]
Swift actually supports inheritance of initializers. For example in a following code Bar will not require initializer, but it will still be available:
```
class Foo {
  let value: Int
  init(value: Int = 5) {
    self.value = value
  }
}

class Bar: Foo {
}
```

The problematic situation will be if we make Foo generic. Somehow in this scenario Swift starts to ask designated initializer override. I do not really understand why yet and what could be an option, but I am trying to find an answer now.

**Comment by ashfurrow** [comment]
Interesting. I suppose what I meant was, a reason to subclass would be to provide custom behaviour, which may need to be given as an initializer parameter Subsequently, you'd want to define all the usual default values and would also need to call the super-initializer. I would love to hear ideas to solve either of those problems – please keep us in the loop about your progress, maybe on a new issue?

**Comment by nikita-leonov** [comment]
While I am going to dig a little bit more, I am not sure that I will be able to handle the whole issue. I suggest to not create a new one, but just reopen current issue. It has whole history of thought, with an actual problem described by you. I think this issue is good enough for tracking everything related to simplifying of subclasses. 

Also I would like to highlight why it is important. As for today we have a lot of code duplication in our product of initializers like following:
```
    override init(endpointClosure: EndpointClosure = ReactiveCocoaMoyaProvider.DefaultEndpointMapping, requestClosure: RequestClosure = ReactiveCocoaMoyaProvider.DefaultRequestMapping, stubClosure: StubClosure = ReactiveCocoaMoyaProvider.NeverStub, manager: Manager = ReactiveCocoaMoyaProvider<IdentificationServiceAPI>.DefaultAlamofireManager(), plugins: [PluginType] = [], stubScheduler: DateSchedulerType? = nil) {
        super.init(endpointClosure: endpointClosure, requestClosure: requestClosure, stubClosure: stubClosure, manager: manager, plugins: plugins, stubScheduler: stubScheduler)
    }
```

With the only difference of manager`s token. Our static analysis tools complaining about this a lot, and I believe it is right complain, as it eventually unnecessary code duplication that hard to read and that does not actually adds value.

**Comment by nikita-leonov** [comment]
Here is a stack overflow question with a problem explained — http://stackoverflow.com/questions/36957858/why-swift-requires-override-of-designated-initializer-of-generic-superclass lets see if someone in community actually can explain this strange generics behavior.

**Comment by nikita-leonov** [comment]
@ashfurrow As result of my research I found some combinations of associated types and default values that actually removes requirement to override designated initializer. It seems that major reason for override requirement in Moya can be some default values. While it is definitely some buggy behavior of Swift, I think it is possible to find workaround even in a current state of Swift.

**Comment by nikita-leonov** [comment]
@ashfurrow good news, an issue that prevents of removing this huge initializers was recently fixed in master — https://bugs.swift.org/browse/SR-416 so this issue prob not bearably will be fixed in Swift 3. 
ps I still suggest to reopen it, as it ongoing and reproducible in current maya with Swift 2.2.
